### ISSUE TITLE
String cgt

### ISSUE THREAD

**Issue by lbj2014** [PowerShell/PowerShell#18638]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

# because of ascii, "A"==65, "a"==97
Write-Host ([char]"A" -cgt [char]"a")  # False
Write-Host ([int][char]"A" -gt [int][char]"a")  # False
# but
Write-Host ("A" -cgt "a")  # True
# can tell me why

### Expected behavior

```console
Write-Host ("A" -cgt "a")
False
```

### Actual behavior

```console
Write-Host ("A" -cgt "a")
True
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.0
PSEdition                      Core
GitCommitId                    7.3.0
OS                             Darwin 21.6.0 Darwin Kernel Version 21.6.0: Wed Aug 10 14:28:23 PDT 2022; root:xnu-8020.141.5~2/RELEASE_ARM64_T6…
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by jhoneill** [comment]
It is more complicated than the ASCII values. 

Consider A  À Å ;  a  à  and å

In English A  À Å all sort together even though US 7-bit ASCII doesn't include accented characters and Unicode must maintain compatibility so the A+Accent isn't near A 

In Norwegian  Å  is treated as a distinct letter. so in English 
```
l> "a", "b" , "y", "å", "z", "Z,", "Å", "Y", "B", "A" | sort-object
a
A
å
Å
b
B
y
Y
z
Z,
```
The accented characters sort on their own, with the unaccented character.  But change the culture... 
``` 
PS> "a", "b" , "y", "å", "z", "Z,", "Å", "Y", "B", "A" | sort-object -Culture no-no
a
A
b
B
y
Y
z
Z,
å
Å
```    

The culture basically dictates if we sort upper before lowercase or lower before upper case, and I think most are upper  

Greater than and less than then work on where the character **would be sorted** , not on its unicode value (because if you we used 8 bit encodings, some characters like € might not have a value or might have different values depending on the active code page. ASCII, being for American in the 1950s doesn't have €) 

 a **sorts before** A so is "less than" even though [int]"a" is greater than [int]"A" .

 b sorts after "À" (but before "B") so 
```
 ps > [int][char]"b" -lt [int][char]"À"
True

ps> "b" -lt "À"
False 
```


.

**Comment by daxian-dbw** [comment]
A `TL;DR` answer that supplement @jhoneill's answer above is:
- for string operands, `-cgt` eventually comes down to the .NET method `Cultureinfo.CompareInfo.Compare("A", "a", CompareOptions.None)`, and the `cultureInfo` used in PoweShell is `InvariantCulture`. The return value is 1 in this case, and that's why `"A" -cgt "a"` returns true.

**Comment by mklement0** [comment]
Let me try a comprehensive summary:

* Terminology:

  * The .NET equivalent of _ASCII values_ - which are limited to the legacy 7-bit [ASCII](https://en.wikipedia.org/wiki/ASCII) "alphabet" (limited to 128 characters) - are _Unicode code points_, because .NET [`[string]`](https://docs.microsoft.com/en-US/dotnet/api/System.String)s are Unicode strings, composed of [`[char]`](https://docs.microsoft.com/en-US/dotnet/api/System.Char) instances, which are UTF-16 code units, i.e. in effect `[uint16]` values, which therefore have a range of `0x0` - `65535` (`0xFFFF`); Unicode's code points are a _superset_ of ASCII values, however.

    * Note that, despite their name, `[char]` instances are only _typically_, but _not always_ whole characters, namely only for those code points in the so-called [BMP (Basic Multilingual Plane)](https://en.wikipedia.org/wiki/Basic_Multilingual_Plane); for characters outside that plane,  i.e. those with a code point _above_ `65535` (`0xFFFF`), which notably includes [emojj](https://en.wikipedia.org/wiki/Emoji), they form either half of a [surrogate pair](https://en.wikipedia.org/wiki/Universal_Character_Set_characters#Surrogates); that is, _two_ `[char]` instances are needed to represent a character outside the BMP.

  * In practice, when strings are compared in a manner that requires _ordering_ (less-than, equal, or greater-than), a linguistic (human-language-based), culture-specific _collation order_ is often used (though the .NET APIs don't use that term) - and such an order is _unrelated_ to (Unicode) code points.

    *  Usually, corresponding lowercase and uppercase letters are grouped together, with lowercase letters coming _before_ their uppercase counterparts, and with accented variants often, but not necessarily, being grouped with their base characters, as @jhoneill's examples show (add `-CaseSensitive` for predictable results).

    * In .NET, these culture-specific rules are implemented via per-culture implementations of the  [`System.Globalization.CompareInfo`](https://learn.microsoft.com/en-US/dotnet/api/System.Globalization.CompareInfo) class.

   * By contrast, comparing strings purely numerically by their characters' _code points_ is called an _ordinal_ comparison.


* PowerShell compares **`[string]` instances** using a **culture-specific collation order**, i.e. it does _not_ perform _ordinal_ comparison.

  * As @daxian-dbw points out, PowerShell's _operators_ use the _invariant_ culture, [`[cultureinfo]::InvariantCulture`](https://docs.microsoft.com/en-US/dotnet/api/System.Globalization.CultureInfo.InvariantCulture)

  * By contrast, [`Sort-Object`](https://learn.microsoft.com/powershell/module/microsoft.powershell.utility/sort-object) uses the _current_ culture (`[cultureinfo]::CurrentCulture`, as also returned by [`Get-Culture`](https://learn.microsoft.com/powershell/module/microsoft.powershell.utility/get-culture)).
  * This discrepancy is unfortunate, and _can_ make a difference, as @jhoneill's examples show (add `-CaseSensitive` for predictable results). Unfortunately, as of v7.3.0, requesting the invariant culture via `-Culture ([cultureinfo]::InvariantCulture)` does _not_ work: see #18655

  * If you want _ordinal_ comparison:
    * For _two given strings_, use `[string]::Compare()` with an ordinal [`System.StringComparison`](https://docs.microsoft.com/en-US/dotnet/api/System.StringComparison) value, e.g.:
      * `[string]::Compare('a', 'A', [StringComparison]::Ordinal)` ... `32`, i.e., the difference in code-point values; that the number is positive implies that `a` is greater than `A`, in an ordinal sense.
      * `[string]::Compare('a', 'A', [StringComparison]::OrdinalIgnoreCase)` ... `0`, because - after culture-invariant transformation of both input strings to uppercase - their code points are identical.
    * For _sorting_, `Sort-Object` does _not_ offer a solution, because the only way to fundamentally modify the sort order is with a `-Culture` argument, and there isn't a culture that that performs _ordinal_ comparisons.
      * However, the [`System.Array`](https://docs.microsoft.com/en-US/dotnet/api/System.Array) type  has  a static [`Sort()`](https://docs.microsoft.com/en-US/dotnet/api/System.Array.Sort) method with an overload that accepts a [`System.Collections.IComparer`](https://docs.microsoft.com/en-US/dotnet/api/System.Collections.IComparer) instance, to which you can pass a [`System.StringComparer`](https://docs.microsoft.com/en-US/dotnet/api/System.StringComparer) instance - note that this method sorts _in place_:
        *  `$a = 'a', 'A'; [Array]::Sort($a, [StringComparer]::Ordinal); $a`  # -> 'A', 'a', i.e. ordinal sorting took place.

* By contrast, **`[char]` instances** always use **(ultimately) _ordinal_ comparison**; note that PowerShell always requires a `[char]` cast, even with single-character string literals such as `'a'`, given that PowerShell has no built-in _literal_ syntax for `[char]` (unlike C#, where `'.<some-char>'` is _exclusively_ a `char` literal):
  * For (non-)_equality_ comparison (`-eq`, `-ne`), _case is respected_, if requested:
    * `[char] 'a' -ceq [char] 'A'` -> $false, because the characters' code points aren't identical.
    * `[char] 'a' -eq [char] 'A'` -> $true, because - after culture-invariant transformation to uppercase, the two characters' code points are identical.
    * `[char] 'a' -ieq [char] 'A'` -> ditto, `-ieq` is effectively an alias of `-eq`
  * For less-than(-or-equal) / greater-than (-or-equal) comparisons, _case is ignored_, because _ordinal_ comparison is _invariably_ performed:
    * `[char] 'a' -cgt [char] 'A'` -> $true, because the code point of char. `a` (LATIN SMALL LETTER A, [`U+0061`](http://www.fileformat.info/info/unicode/char/61)) is greater than the code point of `A` (LATIN CAPITAL LETTER A, [`U+0041`](http://www.fileformat.info/info/unicode/char/41))
    * `[char] 'a' -gt [char] 'A'`-> !! ditto (also with `-igt`); that is, requests to ignore case are _not_ honored.

**Comment by jhoneill** [comment]
> * `[char] 'a' -gt [char] 'A'`-> !! ditto (also with `-igt`); that is, requests to ignore case are _not_ honored.

@mklement0 (a hat tip to your thoroughness, although you must be neglecting other things in life :-)   this looks like a **bug** to me 
```
PS>  [char] 'a' -igt [char] 'A'
True

PS>  [char] 'a' -ieq [char] 'A'
True
```
The above is plain wrong, either it is greater than or it is equal to. Both can't be true. (quantum joke are left as an exercise for the reader) 

I don't know if `char `comparisons should always be case insensitive but **different behaviour between `eq `and `ge/gt/le/lt` must be wrong**.  If culture is addressed (with a "legacy behaviour" option) this should be addressed at the same time. Some subtle bugs could be created where code unwittingly relies on those on current behaviour, so any change probably should be opt in. 

Do we know if this has been flagged before ?

**Comment by mklement0** [comment]
Thanks, @jhoneill; I agree that it is a bug, and I've created #18659

(I've also updated the previous comment, which indeed took me a while to write, but I wanted to have a systematic overview, including for my own notes; I may reuse parts in future Stack Overflow answers)

Note that PowerShell predominantly uses the _invariant_ culture (operators, casts, string interpolation), and, from that vantage point, it is cmdlets that use the _current_ culture (by default) that amount to an inconsistency.
I think `Select-String` is  another example.

Unless there is a good reason for this bifurcation that I am missing.
